### PR TITLE
Removed self paced pl column and added self paced pl ID column

### PR DESCRIPTION
**Pull request by juanmanzojr**
This PR follows the work done in [this PR](https://github.com/code-dot-org/code-dot-org/pull/52648). This PR removes the `self_paced_professional_learning` column and adds a new column `self_paced_pl_course_offering_id` as a foreign key to create an association between a course offering and a self paced professional learning course offering. No data is stored within the `self_paced_professional_learning` column and can be dropped. Edited `course_offering_test.rb` to remove the `self_paced_professional_learning` field so tests can pass.

## Links

- jira ticket: [here](https://codedotorg.atlassian.net/jira/software/c/projects/ACQ/boards/64?modal=detail&selectedIssue=ACQ-653)

## Follow-up work
I will update the test file in a separate PR and serialize using `CourseOffering.all.each(&:write_serialization)`.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/course_offering.rb [hunk 1]
```diff
@@ -19,20 +19,22 @@
...
 #  description                      :string(255)
-#  self_paced_professional_learning :string(255)
 #  professional_learning_program    :string(255)
 #  video                            :string(255)
 #  published_date                   :datetime
+#  self_paced_pl_course_offering_id :integer
 #
 # Indexes
 #
+#  fk_rails_8283d50b87            (self_paced_pl_course_offering_id)
```
Comment (megcrenshaw) [inline dashboard/app/models/course_offering.rb]: Wow! TIL about automatically-generated foreign key names https://guides.rubyonrails.org/v6.0/active_record_migrations.html#foreign-keys

File: dashboard/app/models/course_offering.rb [hunk 2]
```diff
@@ -19,20 +19,22 @@
...
 #  index_course_offerings_on_key  (key) UNIQUE
 #
 
 class CourseOffering < ApplicationRecord
   include Curriculum::SharedCourseConstants
 
   has_many :course_versions, -> {where(content_root_type: ['UnitGroup', 'Unit'])}
+  has_one :self_paced_pl_course_offering, class_name: 'CourseOffering', foreign_key: :self_paced_pl_course_offering_id
```
Comment (megcrenshaw) [inline dashboard/app/models/course_offering.rb]: I just had to read this approximately 17 times, but I think this should be a `belongs_to` association: https://guides.rubyonrails.org/association_basics.html#the-belongs-to-association

File: dashboard/db/migrate/20230720231439_remove_self_paced_column_and_add_self_paced_pl_id_and_add_foreign_key_to_course_offerings.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class RemoveSelfPacedColumnAndAddSelfPacedPlIdAndAddForeignKeyToCourseOfferings < ActiveRecord::Migration[6.1]
+  def change
+    remove_column :course_offerings, :self_paced_professional_learning
+    add_column :course_offerings, :self_paced_pl_course_offering_id, :integer
+    add_foreign_key :course_offerings, :course_offerings, column: :self_paced_pl_course_offering_id
```
Comment (megcrenshaw) [inline dashboard/db/migrate/20230720231439_remove_self_paced_column_and_add_self_paced_pl_id_and_add_foreign_key_to_course_offerings.rb]: It looks like you can instead use `belongs_to` inside the migration instead of `add_foreign_key`, and then you will get *both* the index and the association.

I'm seeing this migration in our repo:
https://github.com/code-dot-org/code-dot-org/blob/c9002ea2294275003c4f51126222603a470b28e3/dashboard/db/migrate/20211026211629_create_code_review_group_members.rb#L3-L8

which leads to
https://github.com/code-dot-org/code-dot-org/blob/2aa1d67265b1bca43bf4277d29c84c60aba20139/dashboard/app/models/code_review_group_member.rb#L10-L22

It looks like since you aren't creating the table, you'd need the slightly-annoying syntax of
https://github.com/code-dot-org/code-dot-org/blob/2aa1d67265b1bca43bf4277d29c84c60aba20139/dashboard/db/migrate/20170522221144_add_course_id_to_sections.rb#L2-L5

**Review by megcrenshaw — COMMENTED** [review]
Juan, this is so complex and it looks great! I think the migration could be slightly simpler by including the `belongs_to` association within the migration itself. But I don't really trust my mental model of Rails things, so @davidsbailey could you take a look at this too?

File: dashboard/app/models/course_offering.rb (hunk 2 above)
Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb]: Ah wait, I'm realizing we're having a model refer to something in the same model ... I have no idea how this works. I'll wait for others to help!

**Comment by juanmanzojr** [comment]
> Juan, this is so complex and it looks great! I think the migration could be slightly simpler by including the `belongs_to` association within the migration itself. But I don't really trust my mental model of Rails things, so @davidsbailey could you take a look at this too?

Sounds good, thank you! Im currently working on a PR that will remove the `self_paced_professional_learning` field from all the serialized files in order to prepare for this migration so any suggestions or changes are appreciated!

File: dashboard/app/models/course_offering.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/course_offering.rb]: I agree with Meg that this should be `belongs_to`.

self-joins are legit and and documented here: https://guides.rubyonrails.org/association_basics.html#self-joins

since the reference name matches the db field name (minus the `_id` part), it might be possible to shorten this to:
```suggestion
  belongs_to :self_paced_pl_course_offering, class_name: 'CourseOffering'
```

However  @juanmanzojr  I think it would be best if you could try this out in the rails console to confirm that it does what you expect before merging.

File: dashboard/app/models/course_offering.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/models/course_offering.rb]: I'm worried that foreign keys might trip us up, since we sometimes take liberties with how we delete and recreate data during seeding. If you are ok with it @juanmanzojr , I would suggest simply not using foreign keys for this model (and any other curriculum-related objects), however I'm very interested in differing opinions here if you think it is important to do.

Comment (davidsbailey) [reply dashboard/app/models/course_offering.rb]: to clarify, I think the index is important, but I am worried we will run into foreign key constraints when deleting objects. I haven't dug into your drone output, but it is possible this could be leading to some of the test failures.

File: dashboard/app/models/course_offering.rb (hunk 2 above)
Comment (juanmanzojr) [reply dashboard/app/models/course_offering.rb]: Ill give this a try! Thank you both!

File: dashboard/app/models/course_offering.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/course_offering.rb]: I had a similar concern, Dave, but I think CourseOffering is okay to use foreign keys (it [uses](https://github.com/code-dot-org/code-dot-org/blob/44784aae5501776ae67433056a30650f92d790cd/dashboard/app/models/course_offering.rb#L373) `find_or_initialize_by`). That said, I had forgotten that we do take liberties with deleting and recreating in other parts of our seeding and agree it might be wise to remove the foreign key, in part for consistency

**Comment by juanmanzojr** [comment]
I have made the suggested changes and tested them in the rails console as well. The association seems to do what we would like it to do.
Screenshot 2023-07-24 at 2 25 16 PM

**Review by davidsbailey — APPROVED** [review]
Looks good!

File: dashboard/app/models/course_offering.rb (hunk 2 above)
Comment (megcrenshaw) [reply dashboard/app/models/course_offering.rb]: Thank you, Dave, for the link to learn about self-joins!
